internal/controller/proposedcommit_controller.go

```diff
@@ -98,33 +106,198 @@ func (r *ProposedCommitReconciler) Reconcile(ctx context.Context, req ctrl.Reque
+	}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ProposedCommitReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&promoterv1alpha1.ProposedCommit{}).
+		Complete(r)
+}
+
+func (r *ProposedCommitReconciler) getGitAuthProvider(ctx context.Context, scmProvider *promoterv1alpha1.ScmProvider, secret *v1.Secret) (scms.GitOperationsProvider, error) {
+	logger := log.FromContext(ctx)
+	switch {
+	case scmProvider.Spec.Fake != nil:
+		logger.V(4).Info("Creating fake git authentication provider")
+		return fake.NewFakeGitAuthenticationProvider(scmProvider, secret), nil
+	case scmProvider.Spec.GitHub != nil:
+		logger.V(4).Info("Creating GitHub git authentication provider")
+		return github.NewGithubGitAuthenticationProvider(scmProvider, secret), nil
+	default:
+		return nil, nil
+	}
+}
+
+func (r *ProposedCommitReconciler) calculateStatus(ctx context.Context, pc *promoterv1alpha1.ProposedCommit, gitOperations *git.GitOperations) error {
+	logger := log.FromContext(ctx)
+
+	branchShas, err := gitOperations.GetBranchShas(ctx, []string{pc.Spec.ActiveBranch, pc.Spec.ProposedBranch})
+	if err != nil {
+		return err
+	}
 	logger.Info("Branch SHAs", "branchShas", branchShas)
 
 	for branch, shas := range branchShas {
 		if branch == pc.Spec.ActiveBranch {
+
 			pc.Status.Active.Hydrated.Sha = shas.Hydrated
-			commitTime, _ := gitOperations.GetShaTime(ctx, shas.Hydrated)
+			commitTime, err := gitOperations.GetShaTime(ctx, shas.Hydrated)
+			if err != nil {
+				return err
+			}
 			pc.Status.Active.Hydrated.CommitTime = commitTime
 
 			pc.Status.Active.Dry.Sha = shas.Dry
-			commitTime, _ = gitOperations.GetShaTime(ctx, shas.Dry)
+			commitTime, err = gitOperations.GetShaTime(ctx, shas.Dry)
+			if err != nil {
+				return err
+			}
 			pc.Status.Active.Dry.CommitTime = commitTime
+
+			activeCommitStatusesState := []promoterv1alpha1.ProposedCommitCommitStatusState{}
+			for _, status := range pc.Spec.ActiveCommitStatuses {
+				var csListActive promoterv1alpha1.CommitStatusList
+				// Find all the replicasets that match the commit status configured name and the sha of the active hydrated commit
+				err := r.List(ctx, &csListActive, &client.ListOptions{
+					LabelSelector: labels.SelectorFromSet(map[string]string{
+						"promoter.argoproj.io/commit-status": utils.KubeSafeLabel(ctx, status.Key),
+					}),
+					FieldSelector: fields.SelectorFromSet(map[string]string{
+						".spec.sha": pc.Status.Active.Hydrated.Sha,
+					}),
+				})
+				if err != nil {
+					return err
+				}
+
+				// We don't want to capture any of the copied commits statuses that are used for GitHub/Provider UI experience only.
+				csListSlice := []promoterv1alpha1.CommitStatus{}
+				for _, item := range csListActive.Items {
+					if item.Labels["promoter.argoproj.io/commit-status-copy"] != "true" {
+						csListSlice = append(csListSlice, item)
+					}
+				}
```

Comment:
Micro-optimization, but could avoid some CPU/mem by traversing this list no more than once

```suggestion
				for _, item := range csListActive.Items {
					if item.Labels["promoter.argoproj.io/commit-status-copy"] == "true" {
						continue
					}
					if found {
						// report TooManyMatching
					}
					found = true
					// do the needful
				}
				if !found {
					// report NoCommitStatusFound
				}
```

Reply:
Don't want to do this quite yet, I think our environment list is always going to be quite slow and I think the short loops read better for now